distribution/lib/Standard/Base/0.0.0-dev/src/Logging.enso

```diff
@@ -21,6 +23,68 @@ Any.log_message self ~message:Text level:Log_Level=Log_Level.Info =
         Log_Level.Severe -> logger.error message
     self
 
+##
+   Infrastructure for reporting logs and progress of computations.
+   Useful for longer running actions (more than a second). A must
+   for long running actions (more than ten seconds). 
+   Start by invoking `Progress.run`, providing your own action
+   to preform and report progress as your action advances. Nest
+   your `Progress.run` invocations freely - the infrastructure is
+   going to take care of combining the overall status of the computations
+   together and will report it as a whole to a
+   [progress displaying facility](https://github.com/enso-org/enso/pull/12126).
+type Progress
+    private Running label:Text max:Float
+
+    ##
+       Performs provided `action` while reporting its progress.
+       `action` is a function that takes a `Progress` handle
+       and uses its methods to report progress of the computation.
+       The value returned from `action` is returned from this
+       method.
+
+       Arguments:
+       - label: Overall description of the action
+       - up_to: how _"long"_ - e.g. how many steps the action is expected to take 
+       - action: function performing the computation while reporting its `Progress`
+
+    run label:Text up_to:Float action:(Progress->Any) -> Any =
+        if up_to>=0.0 . not then Panic.throw "Error"+up_to.to_text
```

Comment:
```suggestion
        if up_to>=0.0 . not then Panic.throw (Illegal_State.Error "The `up_to` value must be non-negative, but was: "+up_to.to_text)
```
(may need an import to actually work)

The error should be more descriptive and an atom. We were discussing about disallowing throwing of non-atom types altogether.

Reply:
Okay I reminded that
> Please review this initial concept.
Sorry if these notes are going unnecessarily deep. Just noting that while I will approve the overall concept, I think we should resolve such things before actually merging the code.

Reply:
Do we have a suitable error already in the stdlib? I don't want to create new type just for this.

Reply:
I suggested `Illegal_State` but it probably should have been `Illegal_Argument`. Both already exist and are used extensively.

Reply:
To make this API resilient against bad values, I decided not to throw any errors in af2ec78c90